pkg/kubefed2/federate/federate.go

```diff
@@ -0,0 +1,268 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package federate
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	apiextv1b1client "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/rest"
+
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis/core/typeconfig"
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	ctlutil "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/kubefed2/options"
+	"github.com/kubernetes-sigs/federation-v2/pkg/kubefed2/util"
+)
+
+var (
+	federate_long = `
+		Federates a Kubernetes resource (including a CRD resource) and
+		enables propagation to all members of federation.
+		Equivalent federated resources of the corresponding kubernetes
+		resource will be created in federation control plane. The kubernetes
+		resource to federate should exist in the cluster hosting the
+		federation control plane. The type of the resource is identified
+		by FEDERATED-TYPE-NAME, which must exist as a FederatedTypeConfig
+		in this Federation Control Plane.
+
+		Current context is assumed to be a Kubernetes cluster hosting
+		the federation control plane. Please use the --host-cluster-context
+		flag otherwise.`
+	// TODO(irfanurrehman): add a flag --resource-cluster-context to pull a resource from another cluster
+
+	federate_example = `
+		# Federate resource named "my-dep" of type identified by FederatedTypeConfig "deployment.apps"
+		kubefed2 federate deployment.apps my-dep --host-cluster-context=cluster1`
+)
+
+type federateResource struct {
+	options.SubcommandOptions
```

Comment:
It contains `dry-run` flag in `options.SubcommandOptions`, but we do not support it in federate a resource.
Do we have plan to do it or it is not necessary for resource case?

Reply:
Updated with dry-run!